homeassistant/components/lifx/__init__.py

```diff
@@ -22,39 +45,173 @@
-
-    hass.data[DOMAIN] = conf or {}
 
-    if conf is not None:
-        hass.async_create_task(
-            hass.config_entries.flow.async_init(
-                DOMAIN, context={"source": config_entries.SOURCE_IMPORT}
-            )
+class LIFXDiscoveryManager:
+    """Manage discovery and migration."""
+
+    def __init__(self, hass: HomeAssistant, migrating: bool) -> None:
+        """Init the manager."""
+        self.hass = hass
+        self.lock = asyncio.Lock()
+        self.migrating = migrating
+        self._cancel_discovery: CALLBACK_TYPE | None = None
+
+    @callback
+    def async_setup_discovery_interval(self) -> None:
+        """Set up discovery at an interval."""
+        if self._cancel_discovery:
+            self._cancel_discovery()
+            self._cancel_discovery = None
+        discovery_interval = (
+            MIGRATION_INTERVAL if self.migrating else DISCOVERY_INTERVAL
         )
+        _LOGGER.debug(
+            "LIFX starting discovery with interval: %s and migrating: %s",
+            discovery_interval,
+            self.migrating,
+        )
+        self._cancel_discovery = async_track_time_interval(
+            self.hass, self.async_discovery, discovery_interval
+        )
+
+    async def async_discovery(self, *_: Any) -> None:
+        """Discovery and migrate LIFX devics."""
+        migrating_was_in_progress = self.migrating
+
+        async with self.lock:
+            discovered = await async_discover_devices(self.hass)
+
+            if legacy_entry := async_get_legacy_entry(self.hass):
+                migration_complete = await async_legacy_migration(
+                    self.hass, legacy_entry, discovered
+                )
+                if migration_complete and migrating_was_in_progress:
+                    self.migrating = False
+                    _LOGGER.debug(
+                        "LIFX migration complete, switching to normal discovery interval: %s",
+                        DISCOVERY_INTERVAL,
+                    )
+                    self.async_setup_discovery_interval()
+
+            if discovered:
+                async_trigger_discovery(self.hass, discovered)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the LIFX component."""
+    hass.data[DOMAIN] = {}
+    migrating = bool(async_get_legacy_entry(hass))
+    discovery_manager = LIFXDiscoveryManager(hass, migrating)
+
+    @callback
+    def _async_delayed_discovery(now: datetime) -> None:
+        """Start an untracked task to discover devices.
+
+        We do not want the discovery task to block startup.
+        """
+        asyncio.create_task(discovery_manager.async_discovery())
+
+    # Let the system settle a bit before starting discovery
+    # to reduce the risk we miss devices because the event
+    # loop is blocked at startup.
+    discovery_manager.async_setup_discovery_interval()
+    async_call_later(hass, DISCOVERY_COOLDOWN, _async_delayed_discovery)
```

Comment:
Why not just wait for `EVENT_HOMEASSISTANT_STARTED` and do discovery then? Would avoid the cooldown need?

Reply:
We want to try to migrate them before the started even if possible so we don't have the least amount of devices that churn after startup